// opensubtitles-client/lib/API.js

var fs = require("fs"),
    xml = require('xml-mapping'),
    Request = require('request'),
    Mustache = require("mustache"),
    Deferred = require('Deferred'),
    utils = require('./Utils.js'),
    events = require( 'events' );


function API(){}

//inherit from EventEmitter
require('util').inherits(API, events.EventEmitter);

API.API_URL = "http://api.opensubtitles.org/xml-rpc";

/*
 * Makes a login request to the API
 *
 * @return{Deferred} with login token as argument
 */
API.prototype.login = function(){

    var scope = this,
        dfd = new Deferred();
    
    scope.emit("beforeLogin");

    utils._getLoginPostData()
        .done(
            function(postData){
                
                utils.request(API.API_URL, postData)
                    .done(
                        function(response){
                            
                            try{
                                var token = utils.parseXmlLoginResult(response);
                                // resolve with token
                                dfd.resolve(token);

                                scope.emit.call(scope, "login", token);
                            }catch(e){
                                scope.emit.call(scope, "error", e);
                            }
                            
                        });
            });
    
    return dfd.promise();
};





/*
 * Makes a search Request
 *
 * @return{Deferred} with login token as argument
 */
API.prototype.search = function(token, lang, query){

    var scope = this,
        dfd = new Deferred();

    // get language
    lang = utils.getOpenSubtitlesLanguage(lang);
    
    utils._getSearchPostData(token, lang, query)
        .done(
            function(postData){
                utils.request(API.API_URL, postData)
                    .done(
                        function(response){
                            
                            try{
                                results = utils.parseXmlSearchResult(response);
                            }catch(e){
                                results = [];
                            }

                            dfd.resolve(results);
                            scope.emit.call(scope, "search", results);
                            
                        });
            });
    
    return dfd.promise();
};

/*
 * Makes a search Request
 *
 * @return{Deferred} with login token as argument
 */
API.prototype.searchForFile = function(token, lang, file){

    // get language
    lang = utils.getOpenSubtitlesLanguage(lang);

    //TODO
    var scope = this,
        dfd = new Deferred(),
        fileSize = 0,
        fileHash = "";
    
    fs.stat(
        file,
        function(err, fd){
            
            fileSize = fd.size;
            
            require("./Hasher.js").getHash(file)
                .done(
                    function(hash){
                        fileHash = hash;
                        
                        utils._getSearchFilePostData(token, lang, fileHash, fileSize)
                            .done(
                                function(postData){
                                    utils.request(API.API_URL, postData)
                                        .done(
                                            function(response){
                                                try{
                                                    results = utils.parseXmlSearchResult(response);
                                                }catch(e){
                                                    results = [];
                                                }

                                                dfd.resolve(results);
                                                scope.emit.call(scope, "search", results);
                                                
                                                
                                            });
                                });
                        
                        
                    }
                );
            
        });
    
    
    return dfd.promise();
    
};


/*
 * Makes a search Request
 *
 * @return{Deferred} when request is made
 */
API.prototype.logout = function(token){

    var dfd = new Deferred();
    
    utils._getLogoutPostData(token)
        .done(
            function(postData){
                utils.request(API.API_URL, postData)
                    .done(
                        function(response){
                           
                            // resolve with token
                            dfd.resolve(response);
                            
                        });
            });
    
    return dfd.promise();
};

/*
 * Get Hash - Powder Implementation
 *
 * @return{Deferred} with login token as argument
 */
API.prototype.getHash = function(file,fileSize){

    //TODO
    var scope = this,
        dfd = new Deferred(),
        fileHash = "";
    
    fs.stat(
        file,
        function(err, fd){
                        
            require("./Hasher.js").getStaticHash(file,fileSize)
                .done(
                    function(hash){
						stopOpenSubtitles = 1;
						console.log("Found a hash: "+hash);
						return hash;
                    }
                );
				            
        });
    
    
    return false;
    
};

//module return new instance of the API
module.exports = new API();
