// mdns-js/test/mdns.test.js
var Lab = require('lab');
var lab = exports.lab = Lab.script();

var describe = lab.describe;
var it = lab.it;
var before = lab.before;
var after = lab.after;
var Code = require('code');   // assertion library
var expect = Code.expect;
var mdns = require('../');



describe('mDNS', function () {
  var browser;
  before(function (done) {
    expect(mdns,  'library does not exist!?').to.exist(mdns);
    browser = mdns.createBrowser();

    browser.on('ready', function onReady(socketcount) {
      expect(socketcount).to.be.above(0);
      done();
    });
  });

  after(function (done) {
    browser.stop();
    done();
  });


  it('should .discover()', function (done) {
    browser.once('update', function onUpdate(data) {
      //mdns._byService.should.have.property('_workstation._tcp');
      expect(data).to.include(['interfaceIndex', 'networkInterface',
        'addresses', 'query']);

      // if (data.query !== '_services._dns-sd._udp.local') {
      //   console.log(data);
      //   data.should.have.property('type');
      // }
      done();
    });
    setTimeout(browser.discover.bind(browser), 500);
  });


  it('should close unused', function (done) {
    browser.closeUnused();
    setTimeout(done, 500);
  });
});
